StenosisMeasurement3D/MRML/vtkMRMLStenosisMeasurement3DParameterNode.h

```diff
@@ -0,0 +1,89 @@
+#include <vtkPolyData.h>
+
+class vtkMRMLMarkupsShapeNode;
+class vtkMRMLMarkupsFiducialNode;
+class vtkMRMLSegmentationNode;
+class vtkMRMLModelNode;
+class vtkMRMLTableNode;
+
+class VTK_SLICER_STENOSISMEASUREMENT3D_MODULE_MRML_EXPORT vtkMRMLStenosisMeasurement3DParameterNode :public vtkMRMLNode
+{
+public:
+    static vtkMRMLStenosisMeasurement3DParameterNode *New();
+    vtkTypeMacro(vtkMRMLStenosisMeasurement3DParameterNode, vtkMRMLNode);
+    void PrintSelf(ostream& os, vtkIndent indent) override;
+    
+    vtkMRMLNode* CreateNodeInstance() override;
+
+    /// Set node attributes from XML attributes
+    void ReadXMLAttributes( const char** atts) override;
+    
+    /// Write this node's information to a MRML file in XML format.
+    void WriteXML(ostream& of, int indent) override;
+    
+    vtkMRMLCopyContentMacro(vtkMRMLStenosisMeasurement3DParameterNode);
+    const char* GetNodeTagName() override {return "StenosisMeasurement3D";}
+
+    void Initialize(); // Reset all member variables to nullptr.
+
+    void SetInputShapeNodeID(const char *nodeID);
+    const char *GetInputShapeNodeID();
+    vtkMRMLMarkupsShapeNode* GetInputShapeNode();
+
+    void SetInputFiducialNodeID(const char *nodeID);
+    const char *GetInputFiducialNodeID();
+    vtkMRMLMarkupsFiducialNode* GetInputFiducialNode();
+
+    void SetInputSegmentationNodeID(const char *nodeID);
+    const char *GetInputSegmentationNodeID();
+    vtkMRMLSegmentationNode* GetInputSegmentationNode();
+
+    vtkSetStdStringFromCharMacro(InputSegmentID);
+    vtkGetCharFromStdStringMacro(InputSegmentID);
+
+    void SetOutputWallModelNodeID(const char *nodeID);
+    const char *GetOutputWallModelNodeID();
+    vtkMRMLModelNode* GetOutputWallModelNode();
+
+    void SetOutputLumenModelNodeID(const char *nodeID);
+    const char *GetOutputLumenModelNodeID();
+    vtkMRMLModelNode* GetOutputLumenModelNode();
+
+    void SetOutputTableNodeID(const char *nodeID);
+    const char *GetOutputTableNodeID();
+    vtkMRMLTableNode* GetOutputTableNode();
+
+    vtkSetObjectMacro(OutputWallOpenPolyData, vtkPolyData);
+    vtkGetObjectMacro(OutputWallOpenPolyData, vtkPolyData)
+
+    vtkSetObjectMacro(OutputLumenOpenPolyData, vtkPolyData);
+    vtkGetObjectMacro(OutputLumenOpenPolyData, vtkPolyData);
+
+    vtkSetObjectMacro(OutputWallClosedPolyData, vtkPolyData);
+    vtkGetObjectMacro(OutputWallClosedPolyData, vtkPolyData);
+
+    vtkSetObjectMacro(OutputLumenClosedPolyData, vtkPolyData);
+    vtkGetObjectMacro(OutputLumenClosedPolyData, vtkPolyData);
+    
+protected:
+    vtkMRMLStenosisMeasurement3DParameterNode();
+    ~vtkMRMLStenosisMeasurement3DParameterNode() override;
+
+    vtkMRMLStenosisMeasurement3DParameterNode(const vtkMRMLStenosisMeasurement3DParameterNode&);
+    void operator=(const vtkMRMLStenosisMeasurement3DParameterNode&);
+
+    std::string InputSegmentID;
+    vtkWeakPointer<vtkPolyData> OutputWallOpenPolyData;
```

Comment:
Changed for vtkSmartPointer.

But vtkGet/SetMacro does not compile.

Reply:
What is the error message? If adding `#include <vtkSmartPointer.h>` does not solve the issue then you can investigate further (copy the error message here, maybe I have an idea). But often we don't need get/set functions, as you can set the variable with a single operator=.

Reply:
Here is a typical error message:

> In file included from /home/user/tmp/Slicer/SlicerExtension-VMTK/StenosisMeasurement3D/MRML/vtkMRMLStenosisMeasurement3DParameterNode.cxx:1:
> /home/user/tmp/Slicer/SlicerExtension-VMTK/StenosisMeasurement3D/MRML/vtkMRMLStenosisMeasurement3DParameterNode.h:62:5: error: no viable overloaded '='
>    62 |     vtkSetMacro(OutputWallOpenPolyData, vtkPolyData);
>       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSetGet.h:134:18: note: expanded from macro 'vtkSetMacro'
>   134 |       this->name = _arg;                                                                           \
>       |       ~~~~~~~~~~ ^ ~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSmartPointer.h:144:20: note: candidate function not viable: no known conversion from 'vtkPolyData' to 'const vtkSmartPointer<vtkPolyData>' for 1st argument
>   144 |   vtkSmartPointer& operator=(const vtkSmartPointer& r)
>       |                    ^         ~~~~~~~~~~~~~~~~~~~~~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSmartPointer.h:151:20: note: candidate template ignored: could not match 'vtkSmartPointer<U>' against 'vtkPolyData'
>   151 |   vtkSmartPointer& operator=(const vtkSmartPointer<U>& r)
>       |                    ^
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSmartPointer.h:165:20: note: candidate template ignored: could not match 'vtkNew<U>' against 'vtkPolyData'
>   165 |   vtkSmartPointer& operator=(const vtkNew<U>& r)
>       |                    ^
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSmartPointer.h:178:20: note: candidate template ignored: could not match 'U *' against 'vtkPolyData'
>   178 |   vtkSmartPointer& operator=(U* r)
>       |                    ^
> In file included from /home/user/tmp/Slicer/SlicerExtension-VMTK/StenosisMeasurement3D/MRML/vtkMRMLStenosisMeasurement3DParameterNode.cxx:1:
> /home/user/tmp/Slicer/SlicerExtension-VMTK/StenosisMeasurement3D/MRML/vtkMRMLStenosisMeasurement3DParameterNode.h:63:5: error: no viable conversion from returned value of type 'vtkSmartPointer<vtkPolyData>' to function return type 'vtkPolyData'
>    63 |     vtkGetMacro(OutputWallOpenPolyData, vtkPolyData)
>       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSetGet.h:156:12: note: expanded from macro 'vtkGetMacro'
>   156 |     return this->name;                                                                             \
>       |            ^~~~~~~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/DataModel/vtkPolyData.h:757:3: note: candidate constructor not viable: no known conversion from 'vtkSmartPointer<vtkPolyData>' to 'const vtkPolyData &' for 1st argument
>   757 |   vtkPolyData(const vtkPolyData&) = delete;
>       |   ^           ~~~~~~~~~~~~~~~~~~
> /home/arc/src/Slicer-SuperBuild9/VTK/Common/Core/vtkSmartPointer.h:197:3: note: candidate function
>   197 |   operator T*() const noexcept { return static_cast<T*>(this->Object); }
>       |   ^
> 5 errors generated.
> 

I'll probably not use the macro.

Reply:
You may need to add `#include <vtkPolyData.h>` as well.

Reply:
`vtkPolyData.h` is already included. `vtkSmartPointer.h` and `vtkSetGet.h` do not help.

I'm using `clang version 18.1.8`, don't know if it's the problem.

Dropping the macro is probably the simplest solution.

Reply:
Grepping through the Slicer code base, it seems that `vtkSet/GetMacro` is always used for data types like int, double... like [here](https://github.com/Slicer/Slicer/blob/98a05fbee8b7eb553ec2ae439e797415e727a984/Libs/vtkSegmentationCore/vtkCalculateOversamplingFactor.h#L80), and `vtkSet/GetObjectMacro` is always used for objects created from high-level classes like [here](https://github.com/Slicer/Slicer/blob/98a05fbee8b7eb553ec2ae439e797415e727a984/Modules/Loadable/Segmentations/Logic/vtkSlicerSegmentationGeometryLogic.h#L102).

Reply:
vtkSet/GetObjectMacro is used when the VTK object is stored as a simple pointer. I'm not sure it is compatible with vtkSmartPointer.

Reply:
Should we then use the like of

`vtkPolyData* GetOutputWallOpenPolyData() {return this->OutputWallOpenPolyData;}`

instead of 

`  vtkGetObjectMacro(OutputWallOpenPolyData, vtkPolyData);`

?

It's an easy rewrite, I just need instructions.